[DOC] Update ingest pipelines content and page navigation (#4285)
opensearch-project/documentation-website: _api-reference/ingest-apis/ingest-pipelines.md

To get started, open **Dev Tools**. From the console, you can create, update, test, and delete your pipelines.

## Create or update an ingest pipeline

Creating an ingest pipeline is a vital step in streamlining your data processing workflow. For example, you can enhance data quality, automate data processing tasks, and ensure your data is prepared and optimized for downstream use. Use create pipeline API to create or update pipelines in OpenSearch.

### Path and HTTP method
```json
PUT _ingest/pipeline/pipeline-id
```

The following is an example request to create a pipeline. 

```json
PUT _ingest/pipeline/pipeline-id
{
  "description": "Optional pipeline description",
  "processors": [
    {
      "set": {
        "description": "Optional processor description",
        "field": "my-long-field",
        "value": 10
      }
    },
    {
      "set": {
        "description": "Set 'my-boolean-field' to true",
        "field": "my-boolean-field",
        "value": true
      }
    },
    {
      "lowercase": {
        "field": "my-keyword-field"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> 

Author: @heemin32 Do we want to include content about if a pipeline fails? Are there common reasons why a pipeline fails?

    },
    {
      "processor_results" : [
        {
          "processor_type" : "set",
          "status" : "success",
          "doc" : {
            "_index" : "index",
            "_id" : "id",
            "_source" : {
              "field-name" : "value",
              "location" : "document-name"
            },
            "_ingest" : {
              "pipeline" : "35678",
              "timestamp" : "2022-02-03T21:45:09.414093212Z"
            }
          }
        }
      ]
    }
  ]
}
```

## Get information about a pipeline

Use the get ingest pipeline API operation to return information about a pipeline.

## Path and HTTP methods

The following example request returns information about a specific pipeline. 

```json
GET _ingest/pipeline/pipeline-id
```
{% include copy-curl.html %}

The following example request returns information about all ingest pipelines.

```
GET _ingest/pipeline
```
{% include copy-curl.html %}

The following is an example response.

```json
{
  "pipeline-id" : {
    "description" : "Optional description of the pipeline",
    "processors" : [
      {
        "set" : {
          "field" : "field-name",
          "value" : "value"
        }
      }

>     ]

Author: @heemin32 Do we want to include a section about creating a pipeline on an existing index?

  }
}
```

## Delete a pipeline

> 

Author: Processor documentation in progress. See [#4193](https://github.com/opensearch-project/documentation-website/issues/4193)

## Create or update an ingest pipeline

Creating an ingest pipeline is a vital step in streamlining your data processing workflow. For example, you can enhance data quality, automate data processing tasks, and ensure your data is prepared and optimized for downstream use. 

Use the create pipeline API operation to create or update pipelines in OpenSearch. Note that the pipeline requires an ingest definition that defines how the processors change the document.

### Path and HTTP method
```json
PUT _ingest/pipeline/pipeline-id
```

### Request body fields

The following table lists the request body fields used to create, or update, a pipeline.

Field | Required | Type | Description
:--- | :--- | :--- | :---
`description` | Optional | String | Description of the ingest pipeline. 
`processors` | Required | Array | The processor that performs a transformation on the documents. Processors run sequentially. 

### Parameters

The following table lists the parameters used with creating, or updating, a pipeline.

Parameter | Required | Type | Description
:--- | :--- | :--- | :---

> `pipeline` | Required | String | The unique identifier, or pipeline id, assigned to the ingest pipeline. A pipeline id is used in API requests to specify which pipeline should be created or modified.  

Reviewer: ```suggestion
`pipeline-id` | Required | String | The unique identifier, or pipeline ID, assigned to the ingest pipeline. A pipeline id is used in API requests to specify which pipeline should be created or modified.  
```

          "status" : "success",
          "doc" : {
            "_index" : "index",
            "_id" : "id",
            "_source" : {
              "field-name" : "value",
              "location" : "document-name"
            },
            "_ingest" : {
              "pipeline" : "35678",
              "timestamp" : "2022-02-03T21:45:09.414093212Z"
            }
          }
        }
      ]
    }
  ]
}
```

## Get information about a pipeline

Use the get ingest pipeline API operation to return information about a pipeline.

## Path and HTTP methods

The following example request returns information about a specific pipeline. 

```json
GET _ingest/pipeline/pipeline-id
```
{% include copy-curl.html %}

The following example request returns information about all ingest pipelines.

```
GET _ingest/pipeline
```
{% include copy-curl.html %}

The following is an example response.

```json
{
  "pipeline-id" : {
    "description" : "Optional description of the pipeline",
    "processors" : [
      {
        "set" : {
          "field" : "field-name",
          "value" : "value"
        }
      }
    ]
  }
}
```

## Delete a pipeline

Use the following requests to delete pipelines. 

> #### Example HTTP methods

Reviewer: Remove this heading

> **Delete a specific pipeline**

Reviewer: ```suggestion
To delete a specific pipeline, pass the pipeline ID as a parameter:
```

```json

> DELETE /_ingest/pipeline/pipeline-id

Reviewer: ```suggestion
DELETE /_ingest/pipeline/<pipeline-id>
```

> DELETE /_ingest/pipeline/pipeline-id

Reviewer: ```suggestion
DELETE /_ingest/pipeline/<pipeline-id>
```

```
{% include copy-curl.html %}

> **Delete all pipelines**

Reviewer: ```suggestion
To delete all pipelines in the cluster, use the wildcard character (`*`):
```

```json
DELETE /_ingest/pipeline/*
```
{% include copy-curl.html %}

> To delete all pipelines in a cluster, the asterisk (*) is necessary.

Reviewer: ```suggestion
```

```json
GET _ingest/pipeline/
```
{% include copy-curl.html %}

### Retrieving information about a specific pipeline

The following example request returns information about a specific pipeline: 

```json
GET _ingest/pipeline/my-pipeline
```
{% include copy-curl.html %}

The response contains the pipeline information:

```json
{
  "my-pipeline": {
    "description": "This pipeline processes student data",
    "processors": [
      {
        "set": {
          "description": "Sets the graduation year to 2023",
          "field": "grad_year",
          "value": 2023
        }
      },
      {
        "set": {
          "description": "Sets graduated to true",
          "field": "graduated",
          "value": true
        }
      },
      {
        "uppercase": {
          "field": "name"
        }
      }
    ]
  }
}
```

## Delete a pipeline

Use the following requests to delete pipelines. 

### Delete a specific pipeline

```json
DELETE /_ingest/pipeline/<pipeline-id>
```
{% include copy-curl.html %}

### Delete all pipelines

```json
DELETE /_ingest/pipeline/*
```
{% include copy-curl.html %}

## Create a pipeline on an existing index

> _<Do we want to include a section on creating a pipeline on an existing index?>

Author: @heemin32 Do we want to include a section about creating a pipeline on an existing index?

Reviewer 2: As long as I know, there is no such thing like creating a pipeline on an existing index.
Once we create a ingest pipeline, we can use the pipeline to add a document to an existing index.

---
layout: default
title: Ingest pipelines
parent: Ingest APIs
nav_order: 10
redirect_from:
  - /opensearch/rest-api/ingest-apis/create-update-ingest/
  - /api-reference/ingest-apis/create-update-ingest/
  - /api-reference/ingest-apis/get-ingest/
  - /api-reference/ingest-apis/simulate-ingest/
  - /api-reference/ingest-apis/delete-ingest/
---

# Ingest pipelines

> Ingest pipelines in OpenSearch are managed by using ingest API operations. When using ingest in production environments, your cluster should contain at least one node with the node roles permission set to `ingest`. For more information about setting up node roles within a cluster, see [Cluster formation]({{site.url}}{{site.baseurl}}/opensearch/cluster/).

Reviewer 2: `ingest` is a role of a node but not permission.

```suggestion
Ingest pipelines in OpenSearch are managed by using ingest API operations. When using ingest in production environments, your cluster should contain at least one node with the `ingest` role. For more information about setting up node roles within a cluster, see [Cluster formation]({{site.url}}{{site.baseurl}}/opensearch/cluster/).
```

Author: Changed in PR #4299

## Prerequisites

The following are prerequisites for using OpenSearch ingest pipelines:

> - <SME insert information>

Author: @heemin32 Please provide any prerquisites details we should include in the documentation. We're aiming to publish this documentation this week. If we need to discuss offline, please let me know. Thank you.

> - <SME insert information>

Reviewer 2: I think this line from description above could be one of prerequisites. `When using ingest in production environments, your cluster should contain at least one node with the `ingest` role.`

In addition to that, following prerequisites is needed when security features are enabled.
```
If the OpenSearch security features are enabled, you must have the `cluster_manage_pipelines` permission to manage ingest pipelines.
```

Author: Revised in PR #4299
